### PR TITLE
Harden session persistence, worker recovery, and daemon refreshes

### PR DESCRIPTION
**Pull request by snimu**
## Summary

- Use atomic writes and zombie-aware lock ownership for durable session and daemon state.
- Keep snapshot errors local to each transfer. Retry recoverable workers and return typed recovery errors.
- Coalesce child-usage accounting while preserving own-session totals after reload. Reuse incremental session scans and passive topology results without caching transient read failures.
- Fix provider billing and request fields, dead-kernel recovery, malformed rendering inputs, and session I/O edge cases.

## Validation

- Node 22: 1,184 coding-agent tests, 51 AI tests, and one TUI test passed. Two existing compaction tests remain skipped.
- Python: 57 runtime tests passed.
- `npm run check` passed, including lint, TypeScript, installer checks, and browser smoke checks.

Refs RES-1260.

---

> [!NOTE]
> **High Risk**
> Touches authentication storage, session transcript repair, daemon protocol/worker recovery, and billing-related provider fields—failures could corrupt sessions, block attach, or mis-price usage.
> 
> **Overview**
> This PR hardens **durable state** and **daemon/session I/O** while fixing a few **provider** and **runtime** edge cases.
> 
> **Persistence and transcripts:** Durable JSON/JSONL writes go through shared **atomic rename** (`writeFileAtomicSync`), including `auth.json` (exclusive create + atomic update), settings, cron jobs, telemetry, and session rewrites. Session files get **crash repair at open** (torn tails, zero-filled lines) when the owning process persists; **export and daemon-client read paths** use in-memory managers so they do not repair another process’s file. The shared **event-log** tail rule is unified (unterminated final line skipped on read, truncated on append). Semantic-edge ledgers now append via that substrate.
> 
> **Daemon performance and correctness:** Session-list metadata uses **incremental resumable scans** with a bounded usage cache and serialized per-path reads. Passive subagent topology is **memoized** until ledger/roster/session inputs change. **Snapshot transfer ids** bind to the materialized event cursor so a mismatched chunked transfer fails locally (clients resync). Failed workers with a live process can be **retried for recovery**; known-but-not-ready sessions return **`session_recovering`** instead of “unknown session”. **Zombie-aware liveness** is shared for `daemon ps`, leases, supervisor ownership, and update-restart. Idle status sweeps **coalesce journal writes** and backoff repeated failed summary generations.
> 
> **Session runtime:** RLM **child usage** flushes one attribution batch per settle boundary (not per child message); live “own usage” subtracts not-yet-indexed child spend. Auto-retry **ends cleanly** when a scheduled `continue()` never starts. Compaction keeps only the **final turn** when the token budget is crossed inside trailing tool results. Bash/output spill paths **degrade** instead of crashing when temp files fail.
> 
> **Providers (`packages/ai`):** OpenAI Responses **omit `service_tier` for GitHub Copilot** (field rejected) but still send explicit tiers elsewhere. Anthropic streaming **reprices cache writes** from `message_delta` `cache_creation` breakdown when usage shifts after `message_start`.
> 
> **Smaller fixes:** Dead IPython kernels drop the memo and restart on next use; piped CLI input joins file/instruction with a blank line; zai default **glm-5.3**; WebP EXIF scan termination; tail truncation and frontmatter BOM handling; daemon protocol **revision 27**.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 1c2c9f2f5db203b4cae94b660f5b9d84c9684fd5. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Harden session persistence with atomic writes, crash repair, and worker recovery
> - Adds shared `writeFileAtomicSync` in [atomic-file.ts](https://github.com/PrimeIntellect-ai/prime-agent/pull/2028/files#diff-b42de982029159020ac68e836a9d6fd23f250e72c2f81120818b88b207738770) and migrates session, auth, event-log, daemon, and harness persistence from bespoke temp-file/rename sequences to the shared utility, preserving symlinks and file modes
> - Adds session-file crash damage detection (`tailLooksDamaged`) and repair (`repairJsonlDamage`) in [session-manager.ts](https://github.com/PrimeIntellect-ai/prime-agent/pull/2028/files#diff-e514ecbe1ca80031cc52600ee3863d993a99616f116355f3fec33c92b532d2cf) that removes zero-filled or torn JSONL tails at open before subsequent appends
> - Replaces size-and-mtime session cache with resumable incremental scanning in `readSessionInfo`/`scanSessionInfo`, so repeated reads resume from the last complete line and concurrent reads for the same path execute in order
> - Reworks daemon worker recovery in [daemon-supervisor.ts](https://github.com/PrimeIntellect-ai/prime-agent/pull/2028/files#diff-18f2d5ec391785bebc6ddf3c1bd3b795e4ed6e02d0aa3852f5bf247d61d012f8) so failed workers with a current process identity can retry recovery, concurrent commands share one in-flight recovery, and snapshot transfer failures trigger client resync only when the failed transfer matches the published snapshot
> - Adds zombie-aware process liveness checks in [child-process.ts](https://github.com/PrimeIntellect-ai/prime-agent/pull/2028/files#diff-a9f6b77c57772609afaa172c89bf3bfb5dd1e9d8ab775b567e40388a0f78fe96) and [daemon-supervisor-ownership.ts](https://github.com/PrimeIntellect-ai/prime-agent/pull/2028/files#diff-6f9ef6ce6a9d2bc51c61fe94d59310efffd26d8f7b3966f5355606b2847d06a5) with a 5-second probe cache, so zombie owners are treated as reclaimable without probing on every fence poll
> - Risk: `readSessionInfo` now serializes same-path reads through a per-path queue and requires a session header as the first parsed entry; files whose first entry is not a header will return `null`. `EventLog.replaySync` now skips any unterminated final line including one that parses as valid JSON, so previously-tolerated partial tails are dropped on replay
>
> 
> #### Changes since #2028 opened
>
> - Introduced read-only session management mode that prevents repairs and writes when loading sessions as a daemon client, during HTML export, or when rendering active daemon summaries [1c2c9f2]
> - Implemented deferred child usage attribution with unindexed usage tracking to ensure own usage reporting excludes child usage that has been accounted live but not yet durably indexed [1c2c9f2]
> - Replaced single-write file initialization with loop-based complete writes that fail fast on short writes for credentials storage and preserved file mode across atomic replacements in harness state saves [1c2c9f2]
> - Removed early-break special case for current file key in session scan cache trimming to enforce retained usage entry cap strictly [1c2c9f2]
> - Added test coverage for read-only session behaviors, unindexed child usage handling, file write safety under partial writes, session scan cache limit enforcement, and permission preservation across atomic saves [1c2c9f2]
> 
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 5ccfad6.</sup>
> 
>

### CODE REVIEW

File: packages/coding-agent/src/core/semantic-edges.ts [hunk 1]
```diff
@@ -375,39 +351,18 @@ export class SemanticEdgeRecorder {
...
+		throw new Error(`corrupt semantic-edge ledger line ${index + 1}: ${String(error)}`);
 	}
-	return { events, validLength };
 }
 
 export function readSemanticEdgeLedger(path: string): SemanticEdgeLedgerEvent[] {
-	return parseLedgerContent(readFileSync(path, "utf8")).events;
+	// A missing ledger stays loud for explicit readers; the recorder treats absence as empty.
+	statSync(path);
+	return new EventLog(path).replaySync(parseSemanticEdgeLine);
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/core/semantic-edges.ts]: Valid — concurrent ledger deletion is a real state (daemon session deletion removes the whole artifact dir, `semantic-edges.jsonl` included, while an external reader may be mid-read), and the `statSync`-then-`replaySync` sequence returned `[]` in that window, masking the pinned ENOENT contract. Fixed in 02f1cdd1d at the premise: the probe is gone and the missing-file decision lives at the single `openSync` — `EventLog.replaySync` takes `missingFileThrows`, so `readSemanticEdgeLedger` propagates ENOENT from the same open that reads (an fd-held file cannot vanish under the read; same structural shape as #1987's bounded-fd fix). The existing pins cover both sides of the contract: missing ledger throws /ENOENT/ without creating the file, and reads never mutate. No stat-then-read path remains in the reader.

File: packages/coding-agent/src/core/event-log.ts [hunk 2]
```diff
@@ -128,57 +124,57 @@ export class EventLog {
 		const payload = [...leadLines, ...lines].join("");
 		const handle = openSync(this.path, "a", 0o600);
 		try {
-			writeSync(handle, payload);
+			// writeSync may write short (e.g. ENOSPC after a prefix); a partial
+			// append reported as success would break write-before-action callers.
+			// TODO(unify): lift to utils/atomic-file writeFullySync when #2035 lands.
+			let buffer = Buffer.from(payload, "utf8");
+			while (buffer.length > 0) {
+				buffer = buffer.subarray(writeSync(handle, buffer));
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/core/event-log.ts]: Valid — the premise check comes out multi-writer: the RLM spawn ledger is multi-process by documented design (the supervisor and each session worker hold their own instance over one file, relying on single-`write()` O_APPEND atomicity; the semantic-edge ledger is single-writer, but the substrate serves both). Completing a short write with a second `write()` could therefore interleave with a rival append and weld two records. Fixed in e88fc9886 with the reclaim shape: on a short write, the appender reads the current tail through its own fd (`a+`) and, only if the tail still equals its torn prefix (it owns the tail), `ftruncate`s the prefix back off — then fails the append either way. A torn tail is read-tolerated and repair-on-append cleans it; a weld is permanent fail-closed corruption; and the failure still satisfies write-before-action (the caller never learns a success). If a rival appended after the torn prefix, nothing is reclaimed and the append fails loudly — that compound (ENOSPC-class short write racing a concurrent append) has no safe completion. Pinned: "reclaims its torn prefix and fails the append on a short write" (injected short write; file byte-identical afterwards, replay clean).

File: packages/coding-agent/src/core/event-log.ts [hunk 3]
```diff
@@ -128,57 +124,57 @@ export class EventLog {
...
+			}
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/core/event-log.ts:133]: Same defect as the High thread on line 132 — resolved together in e88fc9886: the write-fully loop is gone; a short write reclaims its torn prefix (while this writer owns the tail) and fails the append, so a record is never emitted across two write() calls. See the sibling thread for the full multi-writer analysis and the pin.

File: packages/coding-agent/src/core/event-log.ts [hunk 4]
```diff
@@ -126,59 +122,70 @@ export class EventLog {
...
+				// write could interleave with a rival process's append (rlm-ledger is
+				// multi-writer), welding two records. Reclaim the torn prefix while
+				// we still own the tail, then fail: a torn tail is read-tolerated.
+				const size = fstatSync(handle).size;
+				const tail = Buffer.alloc(written);
+				if (
+					readSync(handle, tail, 0, written, size - written) === written &&
+					tail.equals(buffer.subarray(0, written))
+				) {
+					ftruncateSync(handle, size - written);
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/core/event-log.ts]: Valid — the tail-match was not proof of ownership: a rival's committed record whose final bytes coincide with our torn prefix would be truncated, destroying committed data — strictly worse than the torn tail the reclaim tried to prevent. Fixed in d760284d4 by dropping the reclaim entirely (option a): a short write now just fails the append and leaves the torn tail, which is the one tolerated shape — skipped on read, and truncated by any writer's next repair. Verified for both topologies: the single-writer semantic recorder disables on the throw and a resumed recorder's first append runs the repair; every rlm-ledger writer runs repairTailSync before each append, so a rival cleans our prefix. The pin now asserts exactly that lifecycle: short write throws, replay stays clean, and the next append repairs the tail and lands intact. No fd mode change remains (back to plain "a"), no fstat window to reason about, and committed data can never be destroyed.

File: packages/coding-agent/src/core/event-log.ts [hunk 5]
```diff
@@ -126,59 +122,70 @@ export class EventLog {
...
+				}
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/core/event-log.ts]: Valid — the tail-match was not proof of ownership: a rival's committed record whose final bytes coincide with our torn prefix would be truncated, destroying committed data — strictly worse than the torn tail the reclaim tried to prevent. Fixed in d760284d4 by dropping the reclaim entirely (option a): a short write now just fails the append and leaves the torn tail, which is the one tolerated shape — skipped on read, and truncated by any writer's next repair. Verified for both topologies: the single-writer semantic recorder disables on the throw and a resumed recorder's first append runs the repair; every rlm-ledger writer runs repairTailSync before each append, so a rival cleans our prefix. The pin now asserts exactly that lifecycle: short write throws, replay stays clean, and the next append repairs the tail and lands intact. No fd mode change remains (back to plain "a"), no fstat window to reason about, and committed data can never be destroyed.

File: packages/coding-agent/src/core/event-log.ts [hunk 6]
```diff
@@ -128,57 +124,50 @@ export class EventLog {
...
-			const fd = openSync(this.path, "r+");
-			try {
-				const lastByte = Buffer.alloc(1);
-				if (readSync(fd, lastByte, 0, 1, size - 1) !== 1 || lastByte[0] === 0x0a) return;
-				// Truncate guarded by a double-read stability check (cheap
-				// cross-process hardening; a racing append between the check and
-				// the ftruncate stays in the same trust bucket as the documented
-				// O_APPEND small-write atomicity assumption).
-				const first = readAllSync(fd, maxBytes, this.path);
-				const second = readAllSync(fd, maxBytes, this.path);
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/event-log.ts:171]: I’m keeping this fail-closed. If an existing ledger disappears during repair, swallowing `ENOENT` would let `openSync(..., "a")` recreate it without the creation records, because the `onCreate` branch was already bypassed. `SemanticEdgeRecorder` contains the error and continues model calls without provenance headers; the session does not fail. Active-session deletion is also rejected. The existing failed-repair and provenance-disable tests pass.

File: packages/coding-agent/src/core/refinement/refinement.ts [hunk 7]
```diff
@@ -344,17 +335,10 @@ export function mergeHarnessStates(globalState: HarnessState, localState?: Harne
...
-	try {
-		const mode = existsSync(statePath) ? statSync(statePath).mode & 0o777 : 0o600;
-		writeFileSync(tempPath, `${JSON.stringify(state, null, 2)}\n`, { encoding: "utf8", mode });
-		renameSync(tempPath, statePath);
-	} finally {
-		if (existsSync(tempPath)) {
-			unlinkSync(tempPath);
-		}
-	}
+	const targetPath = realpathIfPresentSync(statePath);
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/refinement/refinement.ts:339]: This writer preserves the symlink and writes its target, matching an in-place write. A dangling file target is supported when its parent directory exists; creating missing external backing directories is not part of that contract. The previous writer succeeded here by replacing the symlink itself, not by creating its target directory. The regression check confirms that this case reports `ENOENT` and leaves the alias intact. I’m keeping that behavior.

File: packages/coding-agent/src/utils/child-process.ts [hunk 8]
```diff
@@ -51,6 +51,53 @@ export function isProcessAlive(pid: number): boolean {
...
+/**
+ * Signal the group only while it is provably still the target: the leader process (even a zombie)
+ * anchors its pgid against reuse; once the leader is gone, a live member must hold the pgid at
+ * signal time, narrowing reuse exposure to the inherent kill() TOCTOU of any single-pid signal.
+ */
+export function signalProcessGroupIfHeld(pgid: number, signal: NodeJS.Signals): boolean {
+	if (!processIdExists(pgid) && !processGroupHasLiveMember(pgid)) {
+		return false;
+	}
+	signalProcessGroupOrProcess(pgid, signal);
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/utils/child-process.ts]: `stopTrackedProcess` rechecks the recorded process start identity after shutdown admission and before each TERM/KILL attempt. Its targets include journaled Python kernels without a private process group, so removing the PID fallback would leave those kernels running. A real-process test confirms that requirement, and the leaderless-group tests pass. The remaining check-to-signal race is not made atomic by a group-existence check either. I’m retaining the identity-checked fallback.

File: packages/coding-agent/src/core/agent-session.ts [hunk 9]
```diff
@@ -10712,34 +10791,30 @@ export class AgentSession {
...
+						flushPendingChildUsageAttribution();
 						run.activity = undefined;
 						emitChildUpdate();
 					} else if (event.type === "message_end" && event.message.role === "assistant") {
 						const assistant = event.message as AssistantMessage;
 						if (assistant.stopReason !== "error" && assistant.stopReason !== "aborted") {
+							// Flush before the fold: a persisted aggregate may only include
+							// completions whose childUsage is durable with or before it.
+							flushPendingChildUsageIfStale();
 							attributeChildUsage(parentAssistantForUsage?.usage ?? emptyUsage(), assistant.usage);
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts]: Fixed in 1c2c9f2f5db203b4cae94b660f5b9d84c9684fd5.

Live own-usage reads now exclude pending child attribution for assistant entries in the relevant view, while live aggregate usage remains immediate. The summary memo is invalidated when that accounting changes. Regressions cover branch, fresh and memoized reads while two children settle separately, plus durable reload and scanner totals.

File: packages/coding-agent/src/core/session-manager.ts [hunk 10]
```diff
@@ -927,166 +1037,324 @@ function extractOversizedMessageSummary(line: string): {
...
+}
+
+/** Refresh LRU recency and enforce the retained-usage bound. */
+function storeSessionScanState(filePath: string, state: SessionScanState): void {
+	dropSessionScanState(filePath);
+	state.accountedUsageEntries = state.acc.assistantUsageById.size;
+	retainedUsageEntries += state.accountedUsageEntries;
+	sessionScanStates.set(filePath, state);
+	for (const key of sessionScanStates.keys()) {
+		if (retainedUsageEntries <= SESSION_SCAN_MAX_RETAINED_USAGE_ENTRIES || key === filePath) break;
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/session-manager.ts]: Fixed in 1c2c9f2f5db203b4cae94b660f5b9d84c9684fd5.

The cache now evicts the newest state too when that state alone exceeds the retained-usage limit. The computed listing result is still returned correctly. A 100,001-assistant-entry regression verifies that the next listing rescans rather than retaining the oversized map. This bounds retained cache state, not transient memory needed to scan a large transcript.

File: packages/coding-agent/src/core/session-manager.ts [hunk 11]
```diff
@@ -588,6 +579,125 @@ async function parseEntriesFromBufferAsync(buffer: Buffer): Promise<FileEntry[]>
...
+	const metadata = statMetadataIfPresent(targetPath);
+	const content = keptLines.length > 0 ? `${keptLines.map((kept) => kept.toString("utf8")).join("\n")}\n` : "";
+	try {
+		writeFileAtomicSync(targetPath, content, {
+			...(metadata === undefined ? {} : { mode: metadata.mode }),
+			beforeRename: (tempPath) => {
+				if (metadata !== undefined) chownSync(tempPath, metadata.uid, metadata.gid);
+				// A concurrent appender wins; skipping the repair is safe (next open retries).
+				const current = statSync(targetPath);
+				if (current.size !== snapshot.size || current.mtimeMs !== snapshot.mtimeMs) {
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/session-manager.ts:676]: Fixed in 1c2c9f2f5db203b4cae94b660f5b9d84c9684fd5.

Read-only export, daemon-client startup, and active-summary paths now use non-persistent session managers. They load damaged or older entries without rewriting the backing file; writer opens retain crash recovery. A regression interleaves export with a leased writer repair and append, then verifies the appended entry survives. No additional stat check is treated as atomic compare-and-swap.

File: packages/coding-agent/src/core/agent-session.ts [hunk 12]
```diff
@@ -10599,6 +10636,48 @@ export class AgentSession {
...
+		const flushPendingChildUsageAttribution = () => {
+			if (pendingChildUsageTimer !== undefined) {
+				clearTimeout(pendingChildUsageTimer);
+				pendingChildUsageTimer = undefined;
+			}
+			if (pendingChildUsage.size === 0 || !parentAssistantForUsage) return;
+			const batches = [...pendingChildUsage.entries()];
+			pendingChildUsage.clear();
+			const parentEntry = this._findAssistantEntryForMessage(parentAssistantForUsage);
+			if (!parentEntry) return;
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts]: Fixed in 1c2c9f2f5db203b4cae94b660f5b9d84c9684fd5.

Pending child usage is now retained until the originating assistant is indexed, with a nonblocking flush after the parent event queue drains. Child settlement does not await that queue. The regression uses a real delayed message_end extension and checks attribution, own usage, reload, and scanner results after the child finishes first.

File: prime-agent-runtime/src/rlm/harness.py [hunk 13]
```diff
@@ -295,8 +297,21 @@ def save(self) -> "HarnessState":
...
+        # Atomic replace on the real file: aliases survive, readers never see a torn file.
+        target_path = Path(os.path.realpath(self.file_path))
+        temp_path = target_path.with_name(f"{target_path.name}.{os.getpid()}.{uuid4().hex}.tmp")
+        try:
+            mode = stat.S_IMODE(os.stat(target_path).st_mode)
+        except FileNotFoundError:
+            mode = 0o600
+        try:
+            # The temp carries its final mode from creation: no umask-open window.
+            descriptor = os.open(temp_path, os.O_WRONLY | os.O_CREAT | os.O_EXCL, mode)
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply prime-agent-runtime/src/rlm/harness.py]: Fixed in 1c2c9f2f5db203b4cae94b660f5b9d84c9684fd5.

Existing harness-file permissions are restored on the temporary file with portable path-based chmod before replacement. New files still respect the process umask. The regression covers preserving an existing 0666 mode under umask 0022; the Python harness suite passes.

File: packages/coding-agent/src/core/auth-storage.ts [hunk 14]
```diff
@@ -116,9 +117,21 @@ export class FileAuthStorageBackend implements AuthStorageBackend {
...
+			// Exclusive create: a racing initializer must never replace saved credentials.
+			descriptor = openSync(this.authPath, "wx", 0o600);
+		} catch (error) {
+			if ((error as NodeJS.ErrnoException).code !== "EEXIST") {
+				throw error;
+			}
+			return;
+		}
+		try {
+			writeSync(descriptor, "{}");
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/auth-storage.ts]: Fixed in 1c2c9f2f5db203b4cae94b660f5b9d84c9684fd5.

The initializer retains exclusive wx creation, writes the complete buffer with offset accounting, and throws on zero progress. It never replaces an existing credentials file. Short-write and zero-progress regressions pass.

File: prime-agent-runtime/src/rlm/harness.py [hunk 15]
```diff
@@ -295,8 +297,24 @@ def save(self) -> "HarnessState":
...
+        target_path = Path(os.path.realpath(self.file_path))
+        temp_path = target_path.with_name(f"{target_path.name}.{os.getpid()}.{uuid4().hex}.tmp")
+        try:
+            existing_mode = stat.S_IMODE(os.stat(target_path).st_mode)
+        except FileNotFoundError:
+            existing_mode = None
+        mode = existing_mode if existing_mode is not None else 0o600
+        try:
+            # Create no looser than the destination; retain the umask for new files.
+            descriptor = os.open(temp_path, os.O_WRONLY | os.O_CREAT | os.O_EXCL, mode)
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply prime-agent-runtime/src/rlm/harness.py:310]: Atomic replacement does not preserve arbitrary UID/GID/ACL metadata. This store is per-user or session-local, and the Python kernel inherits the host user’s credentials. The fix preserves existing mode bits and new-file umask. Cross-user ownership and ACL preservation remain outside that guarantee and would require a platform-specific metadata policy.

File: packages/coding-agent/src/core/agent-session.ts [hunk 16]
```diff
@@ -10599,6 +10640,74 @@ export class AgentSession {
...
+						flushPendingChildUsageAttribution(true);
+					};
+					// A message_end extension may still be holding the parent assistant before its append.
+					// The parent drain owns this retry; child settlement never waits for that queue.
+					this._agentEventQueue = this._agentEventQueue.then(flushAfterParentDrain, flushAfterParentDrain);
+					this._agentEventQueue.catch(() => {});
+				}
+				return;
+			}
+			const batches = [...pendingChildUsage.entries()];
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts:10674]: This retains the existing best-effort I/O failure policy: child settlement continues, but a failed append does not advance the durable checkpoint. The real EISDIR regression verifies that behavior. Repeating the append is not idempotent: the original row is already indexed, each call creates another row, and a write error does not prove that no bytes reached disk. A full rewrite can include indexed rows, but an ordinary flush does not guarantee recovery. Reliable retry needs explicit deduplication and index/disk reconciliation; restoring failed origins alone risks duplicate accounting.

**Review by sethkarten — APPROVED** [review]
